[RA1 Ch04] Align with RM 2.4 (#2447)
cntt-n/CNTT: doc/ref_arch/openstack/chapters/chapter04.md

  * [4.3.1 VIM Services](#4.3.1)
  * [4.3.2 Containerised OpenStack Services](#4.3.2)
* [4.4 Consumable Infrastructure Resources and Services](#4.4)

>   * [4.4.1 Support for Cloud Infrastructure Profiles and flavours](#4.4.1)

Author: @joaoarodrigues In Anuket documents we use English (viz. British) spellings. As I pointed out in the write-up "flavor" will be used only when referring to OpenStack flavors.

>   * [4.4.1 Support for Cloud Infrastructure Profiles and flavours](#4.4.1)

Reviewer: This is somewhat confusing. I just left the same comment in PR #2446 
Why not change everywhere to the same spelling?

>   * [4.4.1 Support for Cloud Infrastructure Profiles and flavours](#4.4.1)

Author: @sukhdevkapur The terms have different meanings. OpenStack states that "flavors" are for specify sizing information but then uses it to specify configurations etc. The RM is agnostic of implementations and is "clean" in its use of terminology. In RA1 we need both the RM and OSTK terminology. The use of different spellings, fortuitously for us, gets us around the problem of misuse of the term in OSTK.

>   * [4.4.1 Support for Cloud Infrastructure Profiles and flavours](#4.4.1)

Reviewer 2: @pgoyal01 Flavor is an available hardware configuration for a server. I believe it was originally introduced in Openstack (don't recall to see it used before). 
We agree that RM is agnostic, but then we should remove the flavor term from section/chapter
@sukhdevkapur Agree
@walterkozlowski Lets take this one to discuss in our regular meetings (so we align on the terminology)

>   * [4.4.1 Support for Cloud Infrastructure Profiles and flavours](#4.4.1)

Author: @joaoarodrigues This section is in RA1 and hence "flavor cannot be removed." Both Flavours and Flavor have existed in this chapter -- the addition of this "Note" recently has drawn attention to it.

OpenStack defines flavors as, "In OpenStack, flavors define the compute, memory, and storage capacity of nova computing instances." Later it goes on to say, "Flavors can also determine on which compute host a flavor can be used to launch an instance." The hardware configuration specified  under extra_specs in OpenStack flavors is not what a host is configured as but a subset of configuration information for hosts capable of hosting the "instance". In RM profiles specify how a host is configured.

>   * [4.4.1 Support for Cloud Infrastructure Profiles and flavours](#4.4.1)

Author: @sukhdevkapur @joaoarodrigues There is now only 1 use of "flavour" and that is the reference to RM in the 4.1 Introduction. This conversation should be resolved.

-	Cloud Infrastructure Management Software (VIM: OpenStack) core services and architectural constructs needed to consume and manage the consumable resources
-	Underlying physical compute, storage and networking resources

> This chapter delves deeper into the capabilities of these different resources and their needed configurations to create and operate an OpenStack-based IaaS cloud. This chapter specifies details on the structure of control and user planes, operating systems, hypervisors and BIOS configurations, and architectural details of underlay and overlay networking, and storage, and the distribution of OpenStack service components among nodes. The chapter gets into details into items such as the implementation support for profiles and flavours; the OpenStack flavor types capture both the flavour (sizing) and the profile configuration (of the host).

Author: @joaoarodrigues please do not accept your own suggestions and leave it to the author to do so. Thanks

> This chapter delves deeper into the capabilities of these different resources and their needed configurations to create and operate an OpenStack-based IaaS cloud. This chapter specifies details on the structure of control and user planes, operating systems, hypervisors and BIOS configurations, and architectural details of underlay and overlay networking, and storage, and the distribution of OpenStack service components among nodes. The chapter will also covers implementation support for the [Reference Model profiles and flavours](../../../ref_model/chapters/chapter02.md#24-profiles--flavours); the OpenStack flavor types capture both the sizing and the profile configuration (of the host).

Reviewer 3: ```suggestion
This chapter delves deeper into the capabilities of these different resources and their needed configurations to create and operate an OpenStack-based IaaS cloud. This chapter specifies details on the structure of control and user planes, operating systems, hypervisors and BIOS configurations, and architectural details of underlay and overlay networking, and storage, and the distribution of OpenStack service components among nodes. The chapter also covers implementation support for the [Reference Model profiles and flavours](../../../ref_model/chapters/chapter02.md#24-profiles--flavours); the OpenStack flavor types capture both the sizing and the profile configuration (of the host).
```

#### 4.2.2.5. Compute Nodes

> This section specifies the compute node configurations to support the Basic and High Performance profiles; in OpenStack this would be accomplished by specifying the configurations when creating "flavors". The cloud operator may choose to implement certail profile-extensions ([RM 2.4 Profile Extensions](../../../ref_model/chapters/chapter02.md#242-profile-extensions-specialisations)) as a set of standard configurations, of a given profile, capturing some of the variability through different values or extra specifications.

Reviewer 4: @pgoyal01 typo certail -> certain

Author: @karinesevilla Done - Thanks

-	The software and hardware configurations are as specified in the [Reference Model chapter 5.4](../../../ref_model/chapters/chapter05.md#5.4)
-	BIOS requirement
    -	The general BIOS requirements are described in the [Reference Model chapter 5.4](../../../ref_model/chapters/chapter05.md#5.4)

> **Example Profile Extensions**

Reviewer 4: As the table gathers capabilities related to profile and to extensions, the title can be:
Examples-Profiles and their Extensions

Author: @karinesevilla Done - Thanks

> The Reference Model specifies the Basic (B) and High Performance (H) profile types. The Reference Model also provides a choice of network acceleration capabilities utilising, for example, DPDK and SR-IOV technologies. Table 4-2 lists a few simpe examples of progile extensions and some of their capabilities.

Reviewer 4: Typos:
simpe -> simple
progile -> profile

Author: @karinesevilla Done - Thanks

#### 4.2.2.5. Compute Nodes

> This section specifies the compute node configurations to support the Basic and High Performance profiles; in OpenStack this would be accomplished by creating "flavors". In certain cases, different nodes (hosts, physical servers) in the OpenStack cloud infrastructure may be configured with a profile but with a different geometry (such as CPU over-suscription). The cloud operator may choose to create a "flavor series" as a set of standard configurations, of a given profile, capturing some of this variability.

Reviewer 4: @pgoyal01 
Rather than "flavor series" can we name it "profile series" in order to keep "profiles" for host and flavor for VM?

Author: @karinesevilla Thanks. I think "profile extensions" capture what was the intent originally.

Reviewer 4: Yes okay, resolved

> This section specifies the compute node configurations to support the Basic and High Performance profiles; in OpenStack this would be accomplished by specifying the configurations when creating "flavors". The cloud operator may choose to implement certail profile-extensions ([RM 2.4 Profile Extensions](../../../ref_model/chapters/chapter02.md#242-profile-extensions-specialisations)) as a set of standard configurations, of a given profile, capturing some of the variability through different values or extra specifications.

Author: ```suggestion
This section specifies the compute node configurations to support the Basic and High Performance profiles; in OpenStack this would be accomplished by specifying the configurations when creating "flavors". The cloud operator may choose to implement certain profile-extensions ([RM 2.4 Profile Extensions](../../../ref_model/chapters/chapter02.md#242-profile-extensions-specialisations)) as a set of standard configurations, of a given profile, capturing some of the variability through different values or extra specifications.
```

-	The software and hardware configurations are as specified in the [Reference Model chapter 5.4](../../../ref_model/chapters/chapter05.md#5.4)
-	BIOS requirement
    -	The general BIOS requirements are described in the [Reference Model chapter 5.4](../../../ref_model/chapters/chapter05.md#5.4)

> **Example Profile Extensions**

Author: ```suggestion
**Example Profiles and their Extensions**
```

> The Reference Model specifies the Basic (B) and High Performance (H) profile types. The Reference Model also provides a choice of network acceleration capabilities utilising, for example, DPDK and SR-IOV technologies. Table 4-2 lists a few simpe examples of progile extensions and some of their capabilities.

Author: ```suggestion
The Reference Model specifies the Basic (B) and High Performance (H) profile types. The Reference Model also provides a choice of network acceleration capabilities utilising, for example, DPDK and SR-IOV technologies. Table 4-2 lists a few simple examples of profile extensions and some of their capabilities.
```

Let us consider a compute host with 20 cores with SMT enabled (let us disregard NUMA) and the following parameters specified. The physical cores are numbered '0' to '19' while the sibling threads are numbered '20' to '39' where the vCPUs numbered '0' and '20', '1' and '21', etc. are siblings:

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; cpu_shared_set = 1-7,9-19,21-27,29-39 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; (can also be specified as cpu_shared_set = 1-19,^8,21-39,^28)

This implies that the two physical cores '0' and '8' and their sibling threads '20' and '28' are dedicated to the host services, and 19 cores and their sibling threads are available for Guest instances (and can be over allocated as per the specified cpu_allocation_ratio in nova.conf.

#### 4.2.2.8. Pinned and Unpinned CPUs

> When a VM instance is created the vCPUs are, by default, not assigned to a particular host CPU. Certain compute intensive workloads require real-time or near real-time behavior viz., uninterrupted access to their cores. For such workloads, CPU pinning allows us to bind an instance’s vCPUs to particular host cores or SMT threads. To configure a flavor to use pinned vCPUs, we use a dedicated CPU policy.

Author: @joaoarodrigues "Compute Intensive" is a profile-extension and that is the reason for not using the term. Again, you have accepted your own suggestion and doesn't allow for any discussion.

-	Nodes interfaces segmentation: Have separated NIC ports for Storage and Tenant networks. Actually, the storage traffic is bursty, and especially in case of service restoration after some failure or new service implementation, upgrades, etc. Control and management networks should rely on a separate interface from the interface used to handle tenant networks.
-	Capacity planning: FW, physical links, switches, routers, NIC interfaces and DCGW dimensioning (+ load monitoring: each link within a LAG or a bond shouldn’t be loaded over 50% of its maximum capacity to guaranty service continuity in case of individual failure).
-	Hardware choice: e.g. ToR/fabric switches, DCGW and NIC cards should have appropriate buffering and queuing capacity.

> -	High Performancee compute node tuning (including OVS-DPDK).

Reviewer 4: @goyal Typo: Performancee ->Performance

Author: @karinesevilla Done - Thanks

> -	High Performancee compute node tuning (including OVS-DPDK).

Author: ```suggestion
-	High Performance compute node tuning (including OVS-DPDK).
```

<a name="4.4.1"></a>
### 4.4.1. Support for Cloud Infrastructure Profiles and flavors

Reference Model Chapter 4 and  5 provide information about the Cloud Infrastructure Profiles and their size information. OpenStack flavors with their set of properties describe the VM capabilities and size required to determine the compute host which will run this VM. The set of properties must match compute profiles available in the infrastructure. To implement these profiles and sizes, it is required to set up the flavors as specified in the tables below.

| Flavor Capabilities | Reference<br>RM Chapter 4 and 5 | Basic | High Performance |
|----------|-------------|--------------|-------------|
| CPU allocation ratio (custom extra_specs) | infra.com.cfg.001| In flavor create or flavor set <br>--property cpu_allocation_ratio=4.0 | In flavor create or flavor set <br>--property cpu_allocation_ratio=1.0 |

> | NUMA Awareness | infra.com.cfg.002 | | In flavor create or flavor set specify<br>--property hw:numa_nodes=<integere range of 0 to #numa_nodes – 1><br>To restrict an instance's vCPUs to a single host NUMA node, specify: --property hw:numa_nodes=1<br>Some compute intensive* workloads with highly sensitive memory latency or bandwidth requirements, the instance may benefit from spreading across multiple NUMA nodes: --property hw:numa_nodes=2   |

Reviewer 4: @pgoyal01 typo: integere -> integer

Author: @karinesevilla Done - Thanks

> | NUMA Awareness | infra.com.cfg.002 | | In flavor create or flavor set specify<br>--property hw:numa_nodes=<integere range of 0 to #numa_nodes – 1><br>To restrict an instance's vCPUs to a single host NUMA node, specify: --property hw:numa_nodes=1<br>Some compute intensive* workloads with highly sensitive memory latency or bandwidth requirements, the instance may benefit from spreading across multiple NUMA nodes: --property hw:numa_nodes=2   |

Author: ```suggestion
| NUMA Awareness | infra.com.cfg.002 | | In flavor create or flavor set specify<br>--property hw:numa_nodes=<integer range of 0 to #numa_nodes – 1><br>To restrict an instance's vCPUs to a single host NUMA node, specify: --property hw:numa_nodes=1<br>Some compute intensive* workloads with highly sensitive memory latency or bandwidth requirements, the instance may benefit from spreading across multiple NUMA nodes: --property hw:numa_nodes=2   |
```
